sig-network/plumbing-working-group/[v1] Kubernetes Network Custom Resource Definition De-facto Standard.md

```diff
@@ -0,0 +1,631 @@
+
+```
+{
+    "cniVersion": "0.3.0",
+    "name": "a-bridge-network",
+    "type": "awesome-plugin"
+}
+```
+
+3.4.3 NetworkAttachmentDefinition Object and CNI JSON Configuration Naming Considerations
+
+Per Kubernetes requirements, each `NetworkAttachmentDefinition` object must have a name. The CNI specification strongly recommends that CNI JSON configuration include a name and future CNI specification versions will require a name.
+
+This specification strongly recommends that the `NetworkAttachmentDefinition` object name match the "name" key in CNI JSON configuration referenced by the `NetworkAttachmentDefinition` object (whether stored on-disk or in the NetworkAttachmentDefinition.Spec.Config key). This reduces user confusion and makes the mapping between `NetworkAttachmentDefinition` object and CNI configuration clearer.
+
+While strongly recommended, this matching is not required as it may place Kubernetes API object naming requirements on externally-defined resources. Naming is ultimately left up to the cluster administrator to decide.
+
+**4. Network Attachment Selection Annotation**
+
+To select one or more secondary ("sidecar") networks to which a pod should be attached, this specification defines a Pod object annotation. Because attachments selected by this annotation are secondary attachments, these network attachments are not known to Kubernetes itself and information about them may not be available through the standard Kubernetes API.
+
+The Network Attachment Selection Annotation may be used to select additional attachments to the cluster-wide default network, above and beyond the required initial cluster-wide default network attachment.
+
+4.1 Annotation Name and Format
+
+The Pod object annotation name shall be `k8s.v1.cni.cncf.io/networks`. The annotation value shall be specified in one of two possible formats, described below. Implementations of this specification must support both formats.
+
+Note that even though the objects the annotation references are `NetworkAttachmentDefinition` objects, the annotations name is `k8s.v1.cni.cncf.io/networks`.  This is intentional.
+
+4.1.1 Comma-delimited Format
+
+This format is intended to be a very simple, user-friendly format, composed only of `NetworkAttachmentDefinition` object references separated by commas.  The format of each object reference is either (a) <NetworkAttachmentDefinition object name> to reference a `NetworkAttachmentDefinition` object in the pods namespace, or (b) <NetworkAttachmentDefinition object namespace>/<NetworkAttachmentDefinition object name> to reference a `NetworkAttachmentDefinition` object in a different namespace.
+
+```
+kind: Pod
+metadata:
+  name: my-pod
+  namespace: my-namespace
+  annotations:
+    k8s.v1.cni.cncf.io/networks: net-a,net-b,other-ns/net-c
```

Comment:
Just curious, is `NetworkAttachmentDefinition` have to be namespaced? Since pod can use networks from other namespace.

Reply:
No, `NetworkAttachmentDefinition` can be namespaced, but not have to. If you don't specify namespace, it goes to 'default' namespace. At least, the reference implementation, multus-cni does so.

Reply:
So, `NetworkAttachmentDefinition` it's a namespace-scoped resource?

Reply:
You're right. So you can create same name `NetworkAttachmentDefinition` in different namespaces.

Reply:
@s1061123 by "namespaced" I mean namespace-scoped, cause default namespace also sort of namespaced. : )
Then if another namespace pod use network from default namespace, `k8s.v1.cni.cncf.io/networks` would be `default/net`? 
Is there any isolation that specify network can or cannot be used from different namespace? I think make this namespace-scoped, but without any isolation, only in case same name, I think that's not necessary.

Reply:
@Lion-Wei 'default/net' means 'net' object in 'default' namespace, so you're correct. 
Regarding isolation, I'm not clear what you mean isolation. Do you mean RBAC control or other stuff about isolation?

Reply:
Yeah. . .